File: algorithms/sorting/merge-sort/MergeSort.ts
class Sorting {
	private merge(
		array: number[],
		first: number,
		mid: number,
		last: number,
	): void {
		const newArray: Array<number> = [];
		let i: number = first;
		let j: number = mid + 1;
		const len: number = array.length;

		while (i <= mid && j <= last) {
			if (array[i] < array[j]) {
				newArray.push(array[i]);
				i++;
			} else {
				newArray.push(array[j]);
				j++;
			}
		}

		while (i <= mid) {
			newArray.push(array[i]);
			i++;
		}

		while (j <= last) {
			newArray.push(array[j]);
			j++;
		}

		for (let idx: number = 0; idx < len; idx++) {
			array[idx + first] = newArray[idx];
		}
	}

	public MergeSort(array: number[], first: number, last: number): number[] {
		if (first < last) {
			const mid: number = first + (last - first) / 2;

			this.MergeSort(array, first, mid);
			this.MergeSort(array, mid + 1, last);

			this.merge(array, first, mid, last);
		}

		return array;
	}
}

export { Sorting };

File: algorithms/sorting/merge-sort/MergeSort.dart
class Sorting {
  static _merge(List<int> lst, int first, int mid, int last) {
    List<int> newLst = [];
    int i = first;
    int j = mid + 1;
    int len = lst.length;
    int listIdx = 0;

    while (i <= mid && j <= last) {
      if (lst[i] < lst[j]) {
        newLst.insert(listIdx, lst[i]);
        i++;
      } else {
        newLst.insert(listIdx, lst[j]);
        j++;
      }

      listIdx++;
    }

    while (i <= mid) {
      newLst.insert(listIdx, lst[i]);
      i++;
      listIdx++;
    }

    while (j <= last) {
      newLst.insert(listIdx, lst[j]);
      j++;
      listIdx++;
    }

    for (int index = 0; index < len; index++) {
      lst[index + first] = newLst[index];
    }
  }

  static List<int> MergeSort(List<int> unsortedList, int first, int last) {
    if (first < last) {
      double half = (last - first) / 2;
      int mid = first + half.toInt();

      Sorting.MergeSort(unsortedList, first, mid);
      Sorting.MergeSort(unsortedList, mid + 1, last);

      Sorting._merge(unsortedList, first, mid, last);
    }

    return unsortedList;
  }
}
